ui/ducks/send/send.js

```diff
@@ -0,0 +1,1442 @@
+  INPUT: 'INPUT',
+  MAX: 'MAX',
+};
+
+export const RECIPIENT_SEARCH_MODES = {
+  MY_ACCOUNTS: 'MY_ACCOUNTS',
+  CONTACT_LIST: 'CONTACT_LIST',
+};
+
+async function estimateGasLimitForSend({
+  selectedAddress,
+  value,
+  gasPrice,
+  sendToken,
+  to,
+  data,
+  blockGasLimit = MIN_GAS_LIMIT_HEX,
+}) {
+  // The parameters below will be sent to our background process to estimate
+  // how much gas will be used for a transaction. That background process is
+  // located in tx-gas-utils.js in the transaction controller folder.
+  const paramsForGasEstimate = { from: selectedAddress, value, gasPrice };
+
+  if (sendToken) {
+    if (!to) {
+      // if no to address is provided, we cannot generate the token transfer
+      // hexData, which is the core component to our background process that
+      // estimates gasLimit. We must use our best guess, which is represented
+      // in the gas shared constants.
+      return GAS_LIMITS.BASE_TOKEN_ESTIMATE;
+    }
+    paramsForGasEstimate.value = '0x0';
+    // We have to generate the erc20 contract call to transfer tokens in
+    // order to get a proper estimate for gasLimit.
+    paramsForGasEstimate.data = generateTokenTransferData({
+      toAddress: to,
+      amount: value,
+      sendToken,
+    });
+    paramsForGasEstimate.to = sendToken.address;
+  } else {
+    if (!data) {
+      // eth.getCode will return the compiled smart contract code at the
+      // address if this returns 0x, 0x0 or a nullish value then the address
+      // is an externally owned account (NOT a contract account). For these
+      // types of transactions the gasLimit will always be 21,000 or 0x5208
+      const contractCode = Boolean(to) && (await global.eth.getCode(to));
+      // Geth will return '0x', and ganache-core v2.2.1 will return '0x0'
+      const contractCodeIsEmpty =
+        !contractCode || contractCode === '0x' || contractCode === '0x0';
+      if (contractCodeIsEmpty) {
+        return GAS_LIMITS.SIMPLE;
+      }
+    }
+
+    paramsForGasEstimate.data = data;
+
+    if (to) {
+      paramsForGasEstimate.to = to;
+    }
+
+    if (!value || value === '0') {
+      // ??? Assuming that the value cannot be nullish or 0 to properly
+      // estimate gasLimit?
+      paramsForGasEstimate.value = '0xff';
+    }
+  }
+
+  // If we do not yet have a gasLimit, we must call into our background
+  // process to get an estimate for gasLimit based on known parameters.
+
+  paramsForGasEstimate.gas = addHexPrefix(
+    multiplyCurrencies(blockGasLimit, 0.95, {
+      multiplicandBase: 16,
+      multiplierBase: 10,
+      roundDown: '0',
+      toNumericBase: 'hex',
+    }),
+  );
+  try {
+    // call into the background process that will simulate transaction
+    // execution on the node and return an estimate of gasLimit
+    const estimatedGasLimit = await estimateGas(paramsForGasEstimate);
+    const estimateWithBuffer = addGasBuffer(
+      estimatedGasLimit.toString(16),
```

Comment:
Is this `toString(16)` needed? It doesn't appear to have been there before :thinking:

Reply:
I think this may have been accidentally introduced when dealing with conflicts with https://github.com/MetaMask/metamask-extension/pull/11266

Reply:
https://github.com/MetaMask/metamask-extension/pull/10965/commits/16e66bf469e75fe093950ef106f64c1e064a1138